File: Sources/Pausely/BreakManager.swift
import Foundation
import PauselyCore
import Combine
import SwiftUI

enum BreakStatus {
    case working
    case inBreak
    case paused
}

class BreakManager: ObservableObject {
    static let shared = BreakManager()
    
    @Published var status: BreakStatus = .working
    @Published var timeRemaining: Int = 1200 // Default to 20 minutes (1200 seconds)
    @Published var snoozesLeft: Int = 4
    
    @Published var isSyncedSession: Bool = false
    @Published var isInIntermission: Bool = false
    @Published var intermissionTimeRemaining: Int = 0
    private var anchorTimestamp: TimeInterval = 0
    private var skippedCycleIndices: Set<Int> = []
    private var isApplyingSync = false
    private var snoozeEndTime: Date? = nil
    @Published var currentBreakPrompt: BreakPrompt?
    @Published var currentIntermissionPrompt: MicrobreakPrompt?
    
    private var lastBreakDisplayedTime: Date = Date()
    private var preBreakActiveApp: NSRunningApplication?
    
    private var previousWorkInterval: TimeInterval = 1200
    private var previousBreakDuration: TimeInterval = 20
    
    var anchorTimeString: String {
        guard isSyncedSession else { return "" }
        let date = Date(timeIntervalSince1970: anchorTimestamp)
        let formatter = DateFormatter()
        formatter.timeStyle = .short
        return formatter.string(from: date)
    }
    
    // Configurations
    @Published var workInterval: TimeInterval = AppSettings.shared.workInterval {
        didSet {
            if !isApplyingSync {
                AppSettings.shared.workInterval = workInterval
                isSyncedSession = false
            }
            if status == .working {
                timeRemaining = Int(workInterval)
            }
        }
    }
    @Published var breakDuration: TimeInterval = AppSettings.shared.breakDuration {
        didSet {
            if !isApplyingSync {
                AppSettings.shared.breakDuration = breakDuration
                isSyncedSession = false
            }
            if status == .inBreak {
                timeRemaining = Int(breakDuration)
            }
        }
    }
    
    private var gcdTimer: DispatchSourceTimer?
    private let overlayController = OverlayWindowController.shared
    private var isEnding = false // Guards against repeated endBreak/snooze calls during reverse animation
    private var isScreenLocked = false
    
    /// Direct callback fired at the end of every tick, after all @Published
    /// properties have been set.  Runs on the main queue via GCD — immune to
    /// RunLoop-mode changes (e.g. NSMenu event-tracking).
    var onTick: (() -> Void)?
    
    init() {
        timeRemaining = Int(workInterval)
        startTimer()
        registerScreenLockObservers()
    }
    
    private func registerScreenLockObservers() {
        let dnc = DistributedNotificationCenter.default()
        dnc.addObserver(self,
                        selector: #selector(screenDidLock),
                        name: NSNotification.Name("com.apple.screenIsLocked"),
                        object: nil)
        dnc.addObserver(self,
                        selector: #selector(screenDidUnlock),
                        name: NSNotification.Name("com.apple.screenIsUnlocked"),
                        object: nil)
    }
    
    @objc private func screenDidLock() {
        isScreenLocked = true
        gcdTimer?.suspend()
        CursorWarningManager.shared.hide()
        // Dismiss any active overlay immediately
        if status == .inBreak {
            isEnding = false // allow closeOverlays to work
            overlayController.closeOverlays()
            // Restore working state so the cycle resumes correctly on unlock
            status = .working
            if !isSyncedSession {
                timeRemaining = Int(workInterval)
            }
            snoozesLeft = 4
        }
        // If paused when screen locks, reset to normal working state —
        // locking the screen is essentially a break already.
        if status == .paused {
            status = .working
            timeRemaining = Int(workInterval)
        }
        if isInIntermission {
            isInIntermission = false
            intermissionTimeRemaining = 0
            overlayController.closeOverlays()
        }
    }
    
    @objc private func screenDidUnlock() {
        guard isScreenLocked else { return }
        isScreenLocked = false
        gcdTimer?.resume()
    }
    
    func startTimer() {
        gcdTimer?.cancel()
        let source = DispatchSource.makeTimerSource(queue: .main)
        source.schedule(deadline: .now() + 1.0, repeating: 1.0)
        source.setEventHandler { [weak self] in
            self?.tick()
        }
        source.resume()
        gcdTimer = source
    }
    
    private func tick() {
        // Handle intermission countdown independently
        if isInIntermission {
            intermissionTimeRemaining = TimerTiming.countdown(intermissionTimeRemaining)
            if intermissionTimeRemaining == 0 {
                endIntermission()
            }
        }

        guard !isEnding else { return }
        
        if isSyncedSession {
            let position = TimerTiming.position(work: workInterval, rest: breakDuration,
                                                anchor: anchorTimestamp, now: Date().timeIntervalSince1970)
            let currentCycleIndex = position.cycle
            var newStatus: BreakStatus = position.isBreak ? .inBreak : .working
            var newTimeRemaining = Double(position.remaining)
            if position.isBreak && skippedCycleIndices.contains(currentCycleIndex) {
                newStatus = .working
                newTimeRemaining += workInterval
            }

            // Apply Snooze Override
            if let snoozeEnd = snoozeEndTime {
                if Date() < snoozeEnd {
                    newStatus = .working
                    newTimeRemaining = snoozeEnd.timeIntervalSince(Date())
                } else {
                    snoozeEndTime = nil
                    // Will naturally fall back to math on next tick
                    if newStatus == .inBreak {
                        skippedCycleIndices.insert(currentCycleIndex)
                        newStatus = .working
                        newTimeRemaining = Double(position.remaining) + workInterval
                    }
                }
            }
            
            let oldStatus = self.status
            self.status = newStatus
            self.timeRemaining = Int(ceil(newTimeRemaining))
            
            if oldStatus == .working && newStatus == .inBreak {
                triggerBreak()
            } else if oldStatus == .inBreak && newStatus == .working {
                // If it transitioned naturally to work, we must close overlays
                // We use endBreak to perform the animation gracefully
                endBreak()
            }
            
        } else {
            timeRemaining = TimerTiming.countdown(timeRemaining)
            if timeRemaining == 0 {
                if status == .paused { resumeBreaks() }
                else if status == .working { triggerBreak() }
                else { endBreak() }
            }
        }

        // Pre-fetch wallpaper ~30s before break for instant overlay presentation.
        // This replaces the old 24/7 periodic capture with a targeted on-demand fetch.
        if status == .working && timeRemaining == 30 && !isScreenLocked {
            Task { @MainActor in
                RenderedWallpaperProvider.shared.refresh(for: NSScreen.screens)
            }
        }
        
        // Handle cursor warning logic centrally at the end of tick
        if status == .working && timeRemaining <= 10 && timeRemaining > 0 && !isEnding && !isScreenLocked {
            if CursorWarningManager.shared.viewModel.isVisible == false {
                CursorWarningManager.shared.show(timeRemaining: timeRemaining)
            } else {
                CursorWarningManager.shared.update(timeRemaining: timeRemaining)
            }
        } else if status == .working && timeRemaining > 10 {
            if CursorWarningManager.shared.viewModel.isVisible {
                CursorWarningManager.shared.hide()
            }
        }
        
        onTick?()
    }
    
    func triggerBreak() {
        guard !isEnding else { return }
        
        status = .inBreak
        lastBreakDisplayedTime = Date()
        
        if breakDuration >= 60 {
            currentBreakPrompt = BreakPrompts.randomBreak()
        } else {
            let micro = BreakPrompts.randomMicrobreak()
            currentBreakPrompt = BreakPrompt(title: "Take a breather", message: micro.message)
        }
        
        if !isSyncedSession {
            timeRemaining = Int(breakDuration)
        }
        
        // Hide other apps so the moving wallpaper is visible
        preBreakActiveApp = NSWorkspace.shared.frontmostApplication
        AppVisibilityManager.shared.hideOtherApps()
        
        // Show fullscreen overlay panels across all displays
        overlayController.showOverlays(breakManager: self)
        
        // Play gentle chime
        SoundManager.playStartSound()
    }
    
    func endBreak(wasPremature: Bool = false) {
        guard !isEnding else { return }
        isEnding = true
        
        // Post notification so the overlay can play reverse animations
        NotificationCenter.default.post(name: .breakWillEnd, object: nil)
        
        // Wait for the longest reverse animation (background = 0.50s) + small buffer
        let reverseAnimationDuration = 0.55
        DispatchQueue.main.asyncAfter(deadline: .now() + reverseAnimationDuration) { [weak self] in
            guard let self = self else { return }
            self.isEnding = false
            self.status = .working
            
            if !self.isSyncedSession {
                self.timeRemaining = Int(self.workInterval)
            }
            self.snoozesLeft = 4 // Reset snoozes at full break completion
            
            // Close overlays and warnings
            CursorWarningManager.shared.hide()
            self.overlayController.closeOverlays()
            
            // Restore hidden apps
            AppVisibilityManager.shared.restoreApps()
            
            // Restore previously active app
            self.preBreakActiveApp?.activate(options: .activateIgnoringOtherApps)
            self.preBreakActiveApp = nil
            
            // Play clean finish sound
            if !wasPremature {
                SoundManager.playEndSound()
            }
        }
    }
    
    func snoozeBreak() {
        guard snoozesLeft > 0, !isEnding else { return }
        snoozesLeft -= 1
        isEnding = true
        
        // Post notification so the overlay can play reverse animations
        NotificationCenter.default.post(name: .breakWillEnd, object: nil)
        
        let reverseAnimationDuration = 0.55
        DispatchQueue.main.asyncAfter(deadline: .now() + reverseAnimationDuration) { [weak self] in
            guard let self = self else { return }
            self.isEnding = false
            self.status = .working
            
            let snoozeDuration: TimeInterval = self.workInterval <= 15 ? 10 : 300
            
            if self.isSyncedSession {
                self.snoozeEndTime = Date().addingTimeInterval(snoozeDuration)
            } else {
                self.timeRemaining = Int(snoozeDuration)
            }
            
            // Close overlays and warnings
            CursorWarningManager.shared.hide()
            self.overlayController.closeOverlays()
            
            // Restore hidden apps
            AppVisibilityManager.shared.restoreApps()
            
            // Restore previously active app
            self.preBreakActiveApp?.activate(options: .activateIgnoringOtherApps)
            self.preBreakActiveApp = nil
            
            // No sound when snoozing — silence is intentional
        }
    }
    
    // MARK: - Pause Breaks

    /// The suggested pause duration: 2 full work+break cycles, rounded to
    /// the nearest 15 minutes. For very short test intervals the raw value
    /// is used instead so the rounding doesn't collapse it to zero.
    var suggestedPauseDuration: TimeInterval {
        let raw = 2 * (workInterval + breakDuration)
        let fifteenMinutes: TimeInterval = 900
        let rounded = (raw / fifteenMinutes).rounded() * fifteenMinutes
        return rounded > 0 ? rounded : raw
    }

    func pauseBreaks() {
        guard !isSyncedSession, status == .working else { return }
        CursorWarningManager.shared.hide()
        status = .paused
        timeRemaining = Int(suggestedPauseDuration)
    }

    func resumeBreaks() {
        guard status == .paused else { return }
        status = .working
        timeRemaining = Int(workInterval)
    }

    func skipBreak() {
        if isSyncedSession {
            let cycleDuration = workInterval + breakDuration
            let elapsed = max(0, Date().timeIntervalSince1970 - anchorTimestamp)
            let currentCycleIndex = Int(elapsed / cycleDuration)
            skippedCycleIndices.insert(currentCycleIndex)
        }
        endBreak(wasPremature: true)
    }

    func startIntermission() {
        guard !isInIntermission else { return }
        isInIntermission = true
        intermissionTimeRemaining = Int(breakDuration)
        currentIntermissionPrompt = BreakPrompts.randomMicrobreak()
        
        // Hide other apps so the moving wallpaper is visible
        preBreakActiveApp = NSWorkspace.shared.frontmostApplication
        AppVisibilityManager.shared.hideOtherApps()
        
        // Show fullscreen overlay panels across all displays
        overlayController.showOverlays(breakManager: self, isIntermission: true)
        
        SoundManager.playStartSound()
    }
    
    func endIntermission(wasPremature: Bool = false) {
        isInIntermission = false
        intermissionTimeRemaining = 0
        
        // Post notification so the overlay can play reverse animations
        NotificationCenter.default.post(name: .breakWillEnd, object: nil)
        
        let reverseAnimationDuration = 0.55
        DispatchQueue.main.asyncAfter(deadline: .now() + reverseAnimationDuration) { [weak self] in
            self?.overlayController.closeOverlays()
            AppVisibilityManager.shared.restoreApps()
            self?.preBreakActiveApp?.activate(options: .activateIgnoringOtherApps)
            self?.preBreakActiveApp = nil
            if !wasPremature {
                SoundManager.playEndSound()
            }
        }
    }
    
    func generateSessionCode() -> String {
        if !isSyncedSession {
            previousWorkInterval = workInterval
            previousBreakDuration = breakDuration
            // Backdate the anchor so current timeRemaining is seamless
            if status == .working {
                let elapsed = workInterval - TimeInterval(timeRemaining)
                anchorTimestamp = Date().timeIntervalSince1970 - elapsed
            } else {
                let elapsed = (workInterval + breakDuration) - TimeInterval(timeRemaining)
                anchorTimestamp = Date().timeIntervalSince1970 - elapsed
            }
            anchorTimestamp = floor(anchorTimestamp)
            isSyncedSession = true
        }
        
        return SessionCode.encode(work: Int(workInterval), rest: Int(breakDuration), anchor: anchorTimestamp)
    }

    @discardableResult
    func joinSession(code: String) -> Bool {
        guard let schedule = SessionCode.decode(code, now: Date().timeIntervalSince1970) else { return false }

        if !isSyncedSession {
            previousWorkInterval = self.workInterval
            previousBreakDuration = self.breakDuration
        }
        
        isApplyingSync = true
        self.workInterval = Double(schedule.work)
        self.breakDuration = Double(schedule.rest)
        isApplyingSync = false
        
        self.anchorTimestamp = schedule.anchor
        self.isSyncedSession = true
        self.skippedCycleIndices.removeAll()
        self.snoozeEndTime = nil
        self.snoozesLeft = 4
        
        // If they were in break locally but the code puts them in work, ensure overlays close
        if self.status == .inBreak {
            self.overlayController.closeOverlays()
        }
        status = .working
        if !isScreenLocked { tick() }
        return true
    }
    
    func leaveSession() {
        guard isSyncedSession else { return }
        isSyncedSession = false
        
        isApplyingSync = true
        self.workInterval = previousWorkInterval
        self.breakDuration = previousBreakDuration
        isApplyingSync = false
        
        // Re-anchor based on the last break that was displayed to the user
        let elapsed = Date().timeIntervalSince(lastBreakDisplayedTime)
        if elapsed >= previousWorkInterval {
            // Overdue for a break — trigger immediately
            timeRemaining = 0
        } else {
            timeRemaining = Int(previousWorkInterval - elapsed)
        }
        status = .working
        
        // Clean up sync state
        skippedCycleIndices.removeAll()
        snoozeEndTime = nil
        snoozesLeft = 4
    }
}

File: PauselyWindows/BreakManager.cs
using System;
using PauselyWindows.Services;
using System.Collections.Generic;
using System.Windows.Threading;
using Microsoft.Win32;

namespace PauselyWindows
{
    public class BreakEndedEventArgs : EventArgs
    {
        public bool IsSnoozed { get; }
        public bool PlaySound { get; }
        public BreakEndedEventArgs(bool isSnoozed, bool playSound = true)
        {
            IsSnoozed = isSnoozed;
            PlaySound = playSound;
        }
    }

    public enum BreakStatus
    {
        Working,
        InBreak,
        Paused
    }

    public class BreakManager
    {
        public static BreakManager Shared { get; } = new BreakManager();

        public BreakStatus Status { get; private set; } = BreakStatus.Working;
        public int TimeRemaining { get; private set; } = 1200; // 20 minutes
        public int SnoozesLeft { get; private set; } = 4;
        public BreakPrompt? CurrentBreakPrompt { get; private set; }
        public string? CurrentIntermissionPrompt { get; private set; }

        public bool IsSyncedSession { get; private set; } = false;
        public bool IsInIntermission { get; private set; } = false;
        public int IntermissionTimeRemaining { get; private set; } = 0;
        private double _anchorTimestamp = 0;
        private HashSet<long> _skippedCycleIndices = new HashSet<long>();
        private bool _isApplyingSync = false;
        private DateTime? _snoozeEndTime = null;
        private DateTime _lastBreakDisplayedTime = DateTime.Now;

        private double _previousWorkInterval = 1200;
        private double _previousBreakDuration = 20;

        private double GetCurrentUnixTime() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;

        private double _workInterval = 1200; // 20 minutes
        public double WorkInterval
        {
            get => _workInterval;
            set
            {
                if (!DurationValue.IsValid(value)) return;
                _workInterval = value;
                if (!_isApplyingSync) IsSyncedSession = false;
                if (Status == BreakStatus.Working)
                {
                    TimeRemaining = (int)_workInterval;
                    OnPropertyChanged();
                }
            }
        }

        private double _breakDuration = 20; // 20 seconds
        public double BreakDuration
        {
            get => _breakDuration;
            set
            {
                if (!DurationValue.IsValid(value)) return;
                _breakDuration = value;
                if (!_isApplyingSync) IsSyncedSession = false;
                if (Status == BreakStatus.InBreak)
                {
                    TimeRemaining = (int)_breakDuration;
                    OnPropertyChanged();
                }
            }
        }

        private DispatcherTimer? _timer;
        private bool _isEnding = false;
        private bool _isScreenLocked = false;
        public bool IsScreenLocked => _isScreenLocked;

        public event EventHandler? StatusChanged;
        public event EventHandler? TimerTicked;
        public event EventHandler? BreakTriggered;
        public event EventHandler? BreakEnding;
        public event EventHandler<BreakEndedEventArgs>? BreakEnded;
        public event EventHandler? IntermissionTriggered;
        public event EventHandler<BreakEndedEventArgs>? IntermissionEnded;

        private const double REVERSE_ANIMATION_DURATION_SECONDS = 1.15;
        private const double TEST_MODE_THRESHOLD = 15;
        private const int TEST_MODE_SNOOZE_DURATION = 10;
        private const int STANDARD_SNOOZE_DURATION = 300;

        private BreakManager()
        {
            TimeRemaining = (int)WorkInterval;
            StartTimer();
            SystemEvents.SessionSwitch += SessionSwitch;
        }

        public void StartTimer()
        {
            _timer?.Stop();
            _timer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(1) };
            _timer.Tick += (s, e) => Tick();
            _timer.Start();
        }

        public void StopTimer()
        {
            _timer?.Stop();
        }

        private void Tick()
        {
            // Handle intermission countdown independently.
            if (IsInIntermission)
            {
                IntermissionTimeRemaining = TimerTiming.Countdown(IntermissionTimeRemaining);
                if (IntermissionTimeRemaining == 0) EndIntermission();
                TimerTicked?.Invoke(this, EventArgs.Empty);
            }

            if (_isEnding) return;

            if (IsSyncedSession)
            {
                var position = TimerTiming.At(WorkInterval, BreakDuration, _anchorTimestamp, GetCurrentUnixTime());
                long currentCycleIndex = position.Cycle;
                BreakStatus newStatus = position.IsBreak ? BreakStatus.InBreak : BreakStatus.Working;
                double newTimeRemaining = position.Remaining;
                if (position.IsBreak && _skippedCycleIndices.Contains(currentCycleIndex))
                {
                    newStatus = BreakStatus.Working;
                    newTimeRemaining += WorkInterval;
                }

                if (_snoozeEndTime.HasValue)
                {
                    if (DateTime.Now < _snoozeEndTime.Value)
                    {
                        newStatus = BreakStatus.Working;
                        newTimeRemaining = (_snoozeEndTime.Value - DateTime.Now).TotalSeconds;
                    }
                    else
                    {
                        _snoozeEndTime = null;
                        if (newStatus == BreakStatus.InBreak)
                        {
                            _skippedCycleIndices.Add(currentCycleIndex);
                            newStatus = BreakStatus.Working;
                            newTimeRemaining = position.Remaining + WorkInterval;
                        }
                    }
                }

                BreakStatus oldStatus = Status;
                Status = newStatus;
                TimeRemaining = (int)Math.Ceiling(newTimeRemaining);

                if (oldStatus == BreakStatus.Working && newStatus == BreakStatus.InBreak)
                {
                    TriggerBreak();
                }
                else if (oldStatus == BreakStatus.InBreak && newStatus == BreakStatus.Working)
                {
                    EndBreak();
                }
                else
                {
                    OnPropertyChanged();
                }
            }
            else
            {
                TimeRemaining = TimerTiming.Countdown(TimeRemaining);
                if (TimeRemaining == 0)
                {
                    if (Status == BreakStatus.Paused) ResumeBreaks();
                    else if (Status == BreakStatus.Working) TriggerBreak();
                    else EndBreak();
                }
                TimerTicked?.Invoke(this, EventArgs.Empty);
            }
        }

        public void TriggerBreak()
        {
            if (_isEnding) return;
            Logger.Info("Triggering break.");
            Status = BreakStatus.InBreak;
            _lastBreakDisplayedTime = DateTime.Now;
            CurrentBreakPrompt = BreakDuration >= 60
                ? BreakPrompts.RandomBreak()
                : new BreakPrompt("Take a breather", BreakPrompts.RandomMicrobreak());
            if (!IsSyncedSession)
            {
                TimeRemaining = (int)BreakDuration;
            }
            OnPropertyChanged();
            BreakTriggered?.Invoke(this, EventArgs.Empty);
        }

        public void EndBreak(bool wasPremature = false)
        {
            if (_isEnding) return;
            Logger.Info("Ending break.");
            _isEnding = true;

            BreakEnding?.Invoke(this, EventArgs.Empty);

            var endTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(IsSyncedSession ? Math.Min(REVERSE_ANIMATION_DURATION_SECONDS, WorkInterval / 2) : REVERSE_ANIMATION_DURATION_SECONDS) };
            endTimer.Tick += (s, e) =>
            {
                endTimer.Stop();
                _isEnding = false;
                Status = BreakStatus.Working;
                if (!IsSyncedSession)
                {
                    TimeRemaining = (int)WorkInterval;
                }
                SnoozesLeft = 4;
                OnPropertyChanged();
                BreakEnded?.Invoke(this, new BreakEndedEventArgs(isSnoozed: false, playSound: !wasPremature));
            };
            endTimer.Start();
        }

        public void SnoozeBreak()
        {
            if (SnoozesLeft <= 0 || _isEnding) return;
            Logger.Info($"Snoozing break. Snoozes remaining: {SnoozesLeft - 1}");
            SnoozesLeft -= 1;
            _isEnding = true;

            BreakEnding?.Invoke(this, EventArgs.Empty);

            var endTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(IsSyncedSession ? Math.Min(REVERSE_ANIMATION_DURATION_SECONDS, WorkInterval / 2) : REVERSE_ANIMATION_DURATION_SECONDS) };
            endTimer.Tick += (s, e) =>
            {
                endTimer.Stop();
                _isEnding = false;
                Status = BreakStatus.Working;

                int snoozeDuration = WorkInterval <= TEST_MODE_THRESHOLD ? TEST_MODE_SNOOZE_DURATION : STANDARD_SNOOZE_DURATION;

                if (IsSyncedSession)
                {
                    _snoozeEndTime = DateTime.Now.AddSeconds(snoozeDuration);
                }
                else
                {
                    TimeRemaining = snoozeDuration;
                }

                OnPropertyChanged();
                BreakEnded?.Invoke(this, new BreakEndedEventArgs(isSnoozed: true, playSound: false));
            };
            endTimer.Start();
        }

        public void SkipBreak()
        {
            Logger.Info("Skipping break.");
            if (IsSyncedSession)
            {
                double cycleDuration = WorkInterval + BreakDuration;
                double elapsed = Math.Max(0, GetCurrentUnixTime() - _anchorTimestamp);
                long currentCycleIndex = (long)(elapsed / cycleDuration);
                _skippedCycleIndices.Add(currentCycleIndex);
            }
            EndBreak(wasPremature: true);
        }

        public double SuggestedPauseDuration
        {
            get
            {
                double raw = 2 * (WorkInterval + BreakDuration);
                const double fifteenMinutes = 900;
                double rounded = Math.Round(raw / fifteenMinutes) * fifteenMinutes;
                return rounded > 0 ? rounded : raw;
            }
        }

        public void PauseBreaks()
        {
            if (IsSyncedSession || Status != BreakStatus.Working) return;
            Status = BreakStatus.Paused;
            TimeRemaining = (int)SuggestedPauseDuration;
            OnPropertyChanged();
        }

        public void ResumeBreaks()
        {
            if (Status != BreakStatus.Paused) return;
            Status = BreakStatus.Working;
            TimeRemaining = (int)WorkInterval;
            OnPropertyChanged();
        }

        public void StartIntermission()
        {
            if (IsInIntermission) return;
            Logger.Info("Starting intermission.");
            IsInIntermission = true;
            IntermissionTimeRemaining = (int)BreakDuration;
            CurrentIntermissionPrompt = BreakPrompts.RandomMicrobreak();
            OnPropertyChanged();
            IntermissionTriggered?.Invoke(this, EventArgs.Empty);
        }

        public void EndIntermission(bool wasPremature = false)
        {
            Logger.Info("Ending intermission.");
            IsInIntermission = false;
            IntermissionTimeRemaining = 0;

            BreakEnding?.Invoke(this, EventArgs.Empty);

            var endTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(IsSyncedSession ? Math.Min(REVERSE_ANIMATION_DURATION_SECONDS, WorkInterval / 2) : REVERSE_ANIMATION_DURATION_SECONDS) };
            endTimer.Tick += (s, e) =>
            {
                endTimer.Stop();
                OnPropertyChanged();
                IntermissionEnded?.Invoke(this, new BreakEndedEventArgs(isSnoozed: false, playSound: !wasPremature));
            };
            endTimer.Start();
        }

        public string GenerateSessionCode()
        {
            if (!IsSyncedSession)
            {
                _previousWorkInterval = WorkInterval;
                _previousBreakDuration = BreakDuration;
                if (Status == BreakStatus.Working)
                {
                    double elapsed = WorkInterval - TimeRemaining;
                    _anchorTimestamp = GetCurrentUnixTime() - elapsed;
                }
                else
                {
                    double elapsed = (WorkInterval + BreakDuration) - TimeRemaining;
                    _anchorTimestamp = GetCurrentUnixTime() - elapsed;
                }
                _anchorTimestamp = Math.Floor(_anchorTimestamp);
                IsSyncedSession = true;
            }
            
            return SessionCode.Encode((int)WorkInterval, (int)BreakDuration, _anchorTimestamp);
        }

        public bool JoinSession(string code)
        {
            var schedule = SessionCode.Decode(code, GetCurrentUnixTime());
            if (schedule == null) return false;

            if (!IsSyncedSession)
            {
                _previousWorkInterval = WorkInterval;
                _previousBreakDuration = BreakDuration;
            }

            _isApplyingSync = true;
            WorkInterval = schedule.Work;
            BreakDuration = schedule.Rest;
            _isApplyingSync = false;

            _anchorTimestamp = schedule.Anchor;
            IsSyncedSession = true;
            _skippedCycleIndices.Clear();
            _snoozeEndTime = null;
            SnoozesLeft = 4;

            Logger.Info("Successfully joined sync session.");

            if (Status == BreakStatus.InBreak)
            {
                BreakEnded?.Invoke(this, new BreakEndedEventArgs(isSnoozed: false, playSound: false));
            }
            Status = BreakStatus.Working;
            if (!_isScreenLocked) Tick();
            OnPropertyChanged();
            return true;
        }

        public void LeaveSession()
        {
            if (!IsSyncedSession) return;
            Logger.Info("Leaving sync session and restoring previous intervals.");
            IsSyncedSession = false;

            _isApplyingSync = true;
            WorkInterval = _previousWorkInterval;
            BreakDuration = _previousBreakDuration;
            _isApplyingSync = false;

            // Re-anchor based on the last break that was displayed to the user
            double elapsed = (DateTime.Now - _lastBreakDisplayedTime).TotalSeconds;
            if (elapsed >= _previousWorkInterval)
            {
                // Overdue for a break — trigger immediately
                TimeRemaining = 0;
            }
            else
            {
                TimeRemaining = (int)(_previousWorkInterval - elapsed);
            }
            Status = BreakStatus.Working;

            // Clean up sync state
            _skippedCycleIndices.Clear();
            _snoozeEndTime = null;
            SnoozesLeft = 4;

            Logger.Info($"Restored intervals. WorkInterval: {WorkInterval}, BreakDuration: {BreakDuration}, TimeRemaining: {TimeRemaining}");

            OnPropertyChanged();
        }

        private void OnPropertyChanged()
        {
            StatusChanged?.Invoke(this, EventArgs.Empty);
            TimerTicked?.Invoke(this, EventArgs.Empty);
        }

        private void SessionSwitch(object sender, SessionSwitchEventArgs e)
        {
            Logger.Info($"SessionSwitch event triggered: {e.Reason}");
            System.Windows.Application.Current?.Dispatcher.Invoke(() =>
            {
                if (e.Reason == SessionSwitchReason.SessionLock)
                {
                    Logger.Info("Session locked. Pausing timer and discarding active breaks/intermissions.");
                    _isScreenLocked = true;
                    _timer?.Stop();
                    OnPropertyChanged();

                    if (Status == BreakStatus.InBreak)
                    {
                        _isEnding = false;
                        Status = BreakStatus.Working;
                        if (!IsSyncedSession)
                        {
                            TimeRemaining = (int)WorkInterval;
                        }
                        SnoozesLeft = 4;
                        OnPropertyChanged();
                        BreakEnded?.Invoke(this, new BreakEndedEventArgs(isSnoozed: false, playSound: false));
                    }
                    if (Status == BreakStatus.Paused)
                    {
                        Status = BreakStatus.Working;
                        TimeRemaining = (int)WorkInterval;
                        OnPropertyChanged();
                    }
                    if (IsInIntermission)
                    {
                        IsInIntermission = false;
                        IntermissionTimeRemaining = 0;
                        OnPropertyChanged();
                        IntermissionEnded?.Invoke(this, new BreakEndedEventArgs(isSnoozed: false, playSound: false));
                    }
                }
                else if (e.Reason == SessionSwitchReason.SessionUnlock)
                {
                    Logger.Info("Session unlocked. Resuming timer.");
                    if (_isScreenLocked)
                    {
                        _isScreenLocked = false;
                        _timer?.Start();
                    }
                }
            });
        }
    }
}
